src/FastConv/FastConv.cpp

```diff
@@ -1,32 +1,192 @@
+            // NOTE: We don't need to re-scale the result here, because we already scaled `impulseResponseBlocks` in the constructor.
+            addMultiplySpectra(outputBlock.data(), &impulseResponseBlocks[j*blockLength*2], &inputBlocks[idx*blockLength*2], blockLength);
+        }
+        // Fourier Transform is linear, so we can just do the IFFT once after summing all the spectra.
+        fft->doInvFft(outputBlock.data(), outputBlock.data());
+        // Add back the residue from last time.
+        CVectorFloat::add_I(outputBlock.data(), saved.data(), blockLength);
+        // Move on to next input block.
+        inputBlockIndex = (inputBlockIndex + 1) % numBlocks;
+        // Update index, pointers, and length.
+        indexInBlock = 0;
+        output += remainingInBlock;
+        input += remainingInBlock;
+        length -= remainingInBlock;
+    }
+    // No more full blocks. Check if there are any leftover samples for the next block.
+    if (length > 0) {
+        // Append samples to the next input block.
+        std::copy_n(input, length, &inputBlocks[inputBlockIndex*blockLength*2 + indexInBlock]);
+        // Read out the same number of samples from the last output block.
+        std::copy_n(&outputBlock[indexInBlock], length, output);
+        indexInBlock += length;
+    }
+}
+
+int FreqConvolution::getTailLength() const {
+    // For frequency domain convolution, the tail also includes
+    // the block size due to latency (initial output of zeros).
+    return tailLength;
+}
+
+
+CFastConv::CFastConv() {
+}
+
+CFastConv::~CFastConv() {
+    reset();
+}
+
+Error_t CFastConv::init(float *pfImpulseResponse, int iLengthOfIr, int iBlockLength /*= 8192*/, ConvCompMode_t eCompMode /*= kFreqDomain*/) {
+    if (eCompMode == kTimeDomain) {
+        conv = std::make_unique<TimeConvolution>(pfImpulseResponse, iLengthOfIr);
+    } else if (eCompMode == kFreqDomain) {
+        conv = std::make_unique<FreqConvolution>(pfImpulseResponse, iLengthOfIr, iBlockLength);
+    } else {
+        return Error_t::kFunctionInvalidArgsError;
+    }
     return Error_t::kNoError;
 }
 
-Error_t CFastConv::process (float* pfOutputBuffer, const float *pfInputBuffer, int iLengthOfBuffers )
-{
+Error_t CFastConv::reset() {
+    conv.reset();
     return Error_t::kNoError;
 }
 
-Error_t CFastConv::flushBuffer(float* pfOutputBuffer)
-{
+Error_t CFastConv::process(float *pfOutputBuffer, const float *pfInputBuffer, int iLengthOfBuffers) {
+    if (!conv) return Error_t::kNotInitializedError;
+    conv->process(pfOutputBuffer, pfInputBuffer, iLengthOfBuffers);
     return Error_t::kNoError;
 }
+
+int CFastConv::getTailLength() const {
+    if (!conv) return -1;
+    return conv->getTailLength();
+}
+
+Error_t CFastConv::flushBuffer(float* pfOutputBuffer) {
+    if (!conv) return Error_t::kNotInitializedError;
+    std::vector<float> zeros(getTailLength());
+    return process(pfOutputBuffer, zeros.data(), zeros.size());
```

Comment:
Will 'zeros' be initializes as 0s?

Reply:
Good question! Yes, `vector` constructor (form 4 [here](https://en.cppreference.com/w/cpp/container/vector/vector)) value-initializes its contents - for floats, this defaults to 0.
(This is sort of like doing `float x = float()`; `float x` on its own does not initialize the value, but `float()` returns 0.)

This could be made more explicit by providing a second argument (`std::vector<float> zeros(getTailLength(), 0)`), but the only difference is style - the vector is initialized to zero regardless.